### ISSUE TITLE
detecting motion vectors / handling scrolling better

### ISSUE THREAD

**Issue by totaam** [Xpra-org/xpra#1232]
Scrolling can make us use a video region for the area being scrolled, we should be able to supply the motion vectors to the encoder: we know that the whole region moves as one unit so this should save a lot of CPU and bandwidth.

Plan:
* add optional codec API to pass motion vectors to a test video codec (either x264 or vpx)
* verify that these motion vectors are used: smaller output? metadata? save-to-file feature from [r12144](../commit/0b09f7cc22ce3d75cb5815c7de4b9466cc6cb52b) then playing it back with a player that show motion vectors (ie: [ffdshow](http://www.videohelp.com/software/ffdshow) does)
* compare with and the current search algorithms (try EXA, HEX, etc..) and find their limitations - unit test it all
* write simple motion detection for video region (exact match only and vertical only, for now) - unit test it, see [Search Algorithms for Block-Matching in Motion Estimation](http://www.ece.cmu.edu/~ee899/project/deepak_mid.htm)
* pass the motion vector to the encoder (and turn off deblocking?)
* provide a DBUS api for passing those motion vectors from the app

Alternatively, maybe we could also use the motion vector internally: send the delta (using a fast stream compressor like lz4) + motion: still send the newly exposed region separately (different encoding)
We may want to force an I frame for the next video frame since the delta will be bigger.

Also worth thinking about: distinguish video from scrolling. Video has high framerate and lasts longer, we can then:
* use a different motion search algorithm: use ESA (aka exhaustive search) for video, for scrolling, we should stick with HEX, DIA or UHM (see [X264 - Motion Estimation Method- Comparison](http://mmmash.blogspot.com/2013/03/x264-motion-estimation-method-comparison.html))
* disable "subme" (subpixel motion) for scrolling
* disable partitions?
* only use a single reference frame for scrolling (much faster)

There are probably some limitations on motion vectors range, etc

**Comment by totaam** [comment]
The good news first: with the patch above, I can detect scrolling in the majority of cases where the distance is small enough, ie: with Firefox:
```
> best scroll guess, matches 91% of 464 lines: 10
> best scroll guess, matches 89% of 464 lines: 13
> best scroll guess, matches 88% of 464 lines: 15
> best scroll guess, matches 95% of 464 lines: 6
> best scroll guess, matches 93% of 464 lines: 8
> best scroll guess, matches 89% of 464 lines: 13
> best scroll guess, matches 95% of 464 lines: 6
> best scroll guess, matches 90% of 464 lines: 12
> best scroll guess, matches 95% of 464 lines: 5
> best scroll guess, matches 96% of 464 lines: 5
> best scroll guess, matches 97% of 464 lines: 3
> best scroll guess, matches 95% of 464 lines: 7
> best scroll guess, matches 97% of 464 lines: 3
> best scroll guess, matches 98% of 464 lines: 0
> best scroll guess, matches 90% of 464 lines: -14
> best scroll guess, matches 73% of 464 lines: -42
> best scroll guess, matches 64% of 464 lines: -52
> best scroll guess, matches 74% of 464 lines: -32
> best scroll guess, matches 73% of 464 lines: -32
> best scroll guess, matches 52% of 464 lines: -60
> best scroll guess, matches 77% of 464 lines: -28
> best scroll guess, matches 64% of 464 lines: -45
> best scroll guess, matches 85% of 464 lines: -17
> best scroll guess, matches 85% of 464 lines: -18
```
The bad news is that x264 does not support supplying motion vectors: [Manually feeding x264 with my own motion data?](http://stackoverflow.com/questions/6045116/): *Short answer: No you can't feed in your motion estimation data to x264.*
NVENC does, but it's very poorly documented (as usual) and it is per block... and blocks are small.
 
So I think we'll just have to handle scrolling without using video, unlike the current test patch which hooks in the video path. But we still want to only fire it when there is a good chance it will be useful (as we have to scan the whole picture, and then compare all the checksums)

**Comment by totaam** [comment]
More preparatory work in [r12881](../commit/9f330058993103786e5524b5c184df57176d00e1) + [r12885](../commit/682053d584dd6b536af03e51fbb098f0d676d4c8) + [r12886](../commit/8865956ff967933383bce7539c2ee68c4c95a589) + [r12887](../commit/28e0e9db88ff1ecfde4bddc5c32b7d0119278f3f), and tests in [r12880](../commit/4e2b3980265a34e85022f10f60b1015c11bb6292) + [r12882](../commit/4709cc7c05a295278a5f093e49cbb2f127796861) + [r12883](../commit/78e3bddf0b762e01a2eefe00a92d1e685c2e4afd) + [r12884](../commit/6cafc0d3d1bcea5c5df9a66b4d976e13fc83c328)

**Comment by totaam** [comment]
Lots of changes related to scrolling have been merged:
* [r12912](../commit/bd292e6e8e07ace78f7c467ee2257873ba8db99d) server side support
* [r12914](../commit/85583034060caaf93b850f074d2a87ae083e99a9) makes it possible to record and replay scrolling data
* [r12913](../commit/4712f8a380d9c3f287cdec36d30809cc7738d65c) tests for the new motion class
* [r12901](../commit/ec6154b1e3de9caae348870bca83db89a6f33e8e) sets `XPRA_SCROLL_ENCODING=0` by default client-side since the gl backend has problems with more than one window..
* [r12894](../commit/52400f3409455f5338617aa69d160bbb39dac2d0) adds horizontal scrolling support in client backends (not supported by the server as detection would be much more costly)
* [r12893](../commit/415cec54fca9fe2ad464336811082d9a88d8b630) client backend work
* [r12897](../commit/94018551efa5ea568c5209b633a5244ef3a9e0f7) + [r12896](../commit/af44b7d54be413cfa8ee5b4c2ccb45da884b124b) tests improvements
* [r12909](../commit/7906ea673b9b2974a1c5a5e192a8c38f8d634a8b) ensures that a refresh will also match the window size (even when the video encoder would round the dimensions down to an even number - as is the case with an xterm)


----

Still TODO:
* fix opengl backing with multiple windows (and fix pixmap backing?)
* refresh handling: maybe do something simple, like extending the current refresh regions by the maximum scroll distances
* move scroll handling out of the video-only path, and make it work with the video region code
* check for issues with desktop scaling or transparency
* support for scrolling in the html5 client
* non-scroll packets bypass too much of the send accounting logic: maybe return a list of packets from the encode function?
* fix issues with b-frames #800

Maybes (see comment:2 for details):
* reset the video encoder after scrolling
* move scrolling outside video path
* handle refresh updates "correctly", moving and splitting the list of damaged rectangles - hard
* crc would be quicker if we did it whilst copying the image (when we call restride)
* sub-images don't need to copy the pixels (use reference counting)
* opengl: don't keep temporary fbo around for nothing, use the same code to tackle #478
* merge non-scroll regions separated by small gaps
* interpolated smooth scrolling

**Comment by totaam** [comment]
* [r12920](../commit/b495f4c32470cb473107524e3b6f3aacb99f55c8) finally fixes the opengl client, and scrolling is now enabled by default, can still be disabled by setting `XPRA_SCROLL_ENCODING=0`
* [r12921](../commit/6ef06ad29db5e34d0cb0406f54515663a236f6f4) allows us to override the default `XPRA_SCROLL_MIN_PERCENT=40`, below this percentage, we won't be using scroll encoding, defaulting back to video
* [r12928](../commit/698a6a99c3ec25d05f8f478b10f49b1e0a801fde) was missing from previous commits, you need this to build the new cython "motion" code


----

Until this is integrated more tightly, you need to run the server with `XPRA_ENCODING_STRICT_MODE=1` to trigger the video code path, which is the place where the scrolling detection is currently hooked up.
You can get the scrolling debug with "-d scroll".
The color shown with opengl paint box debugging is orange, it only draws around the area we copied to (not from).

**Comment by totaam** [comment]
[r13402](../commit/6929cdec831c46b22452edcf275d2d36b723a487) adds an SSE 4.2 accelerated CRC32 version which is a lot faster than the standard zlib.crc32 on supported hardware (tested on Skylake).

**Comment by totaam** [comment]
There was a silly bug in the code, which means that the performance was overestimated for crc32c. But as of [r13405](../commit/1c56af3d607458ea75ea1a3e1f9ebdac00b71a0b) we switch to [xxhash](https://github.com/Cyan4973/xxHash) via [python-xxhash](https://pypi.python.org/pypi/xxhash/), and we get 7.5GB/s! (with some better tests)

**Comment by totaam** [comment]
Lots of updates: [r13433](../commit/612c6cf4a93a1008bcc57ea7e48d28678bef1bc0), [r13434](../commit/c425947dfdccfc1ad8bc77697b658e335cf0064c), [r13435](../commit/65342866ca2b269f72d84e2a824a0d4e19c1315b), [r13436](../commit/e38aecd5ee386eb651807f3fe8652cedca542775).
scrolling detection is enabled again, and seems to work well enough. As of [r13442](../commit/5e140af3fd297286386af4848b07590069c0c4c2) we only enable scrolling detection if "xxhash" is available (the alternatives are just too slow).
Still todo: #1257, try to manage conflicts between video regions and scrolling... very difficult to have both at the same time.

**Comment by totaam** [comment]
Looks ready for a first round of testing.
How I tested: using firefox and scrolling around (without video playing at the same time). You can try to see what encoding is used with opengl paint debug ("Meta+Shift+F12" to turn on for a window), but since the scrolling will accumulate previous paint boxes... it can get messy quickly. A better way is to check what encodings are actually being used with `xpra info | egrep "total_frames|total_pixels"`.
Then for extra verbose details, there's "-d scroll".

Important: you must have python-xxhash installed on the server for scroll detection.

**Comment by totaam** [comment]
FYI: the win32 and osx shadow servers now take advantage of scrolling detection to save a huge amount of bandwidth and CPU, they can be useful for testing and helped uncover an absolute monster of a bug: 13479

**Comment by totaam** [comment]
* support for exclusion zones has been added, see #1295.
* adding python-xxhash package for debian server support done in [r13496](../commit/431fdd03f11ce100afb91a2cdc9e87486267ad61)

**Comment by totaam** [comment]
Just hit this:
```
Traceback (most recent call last):
  File "/usr/lib64/python2.7/site-packages/xpra/client/gl/gl_window_backing_base.py", line 514, in do_scroll_paints
    assert y+ydelta>=0 and y+h+ydelta<=bh, "vertical scroll by %i: rectangle %s overflows the backing buffer size %s" % (ydelta, (x, y, w, h), self.size)
AssertionError: vertical scroll by 299: rectangle (17, 2, 1002, 299) overflows the backing buffer size (1021, 591)
```
Not sure how to reproduce it.

**Comment by totaam** [comment]
The bug in comment:14 is caused by a mismatch between the server window size and the backing client-side (which can happen more easily with desktop-scaling enabled, ie: #1098), [r13618](../commit/ff3aff8ed6774a86e55b88a9cd3512b2dc969d1a) is a workaround for that.
[r13617](../commit/4c7b239c73a9e5685e2c6ec46f26d0e675fd73bb) also improves error handling end reporting.

**Comment by totaam** [comment]
Minor change: as of [r13797](../commit/99ec1d9b0182a6dcbc8e551e75c63ef2655752b7) + [r13799](../commit/4f6bfaf46701eb3be0dfb8dcf667ef6664c88b4c), we run the scrolling detection in more cases, even when we detected a video region that really does look like real video, but not if the API has been used to set this video region.

Notes:
* we don't handle auto-refresh as well as we could, see #1320.
* in the future we may be able to re-use this code for shadow servers: #1321 for osx
* as of [r13775](../commit/0cf3bf824b72529ec45a1e68db208a9023ebb5c0), we can switch to and from "true" video on the fly, see #800#comment:25
* as of [r13980](../commit/87e175f8359d9caf830763b01b39b62ff1c6455f): the required python-xxhash package has been added as a dependency and should now get installed by default (#1331)

**Comment by totaam** [comment]
The zero division error should be fixed in [r14053](../commit/dd8d6b8a5195330e8dadfa7eb07894be7f514ec8).

* setting `XPRA_SCROLL_ENCODING` is no longer needed, it has been enabled by default since [r13436](../commit/e38aecd5ee386eb651807f3fe8652cedca542775)
* setting `XPRA_ENCODING_STRICT_MODE` is not needed, and definitely not recommended as this will disable scrolling detection now

----

Scrolling should work out of the box, without needing any special command line arguments, provided that:
* the required xxhash library is installed on the server
* the client uses the opengl accelerated backend

----

To verify that scrolling is being used:
* for each window forwarded, the "scrolling" flag should be on:
```shell
$ xpra info :2 | grep -i scrolling
window.1.encoding.scrolling=True
```
* "-d scroll" on the server will always show details when scrolling detection is firing (ie: when scrolling, holding the enter key down in an xterm is also enough to trigger it), ie:
```
updated scroll data
best scroll guess took 1ms, matches 96% of 312 lines: 0
encode_scrolling(XShmImageWrapper(BGRX: 17, 2, 480, 312), {..}, [], [], {'av-sync': True}) window-dimensions=(499, 316)
 will send scroll packets for yscroll=[0, 13, -13, 13, -13, 26, -26, 26, -26, 39, -39, 39, -39, 52, -52, 52, -52, -65, 65, -65, 65, -78, 78, -78, 78, -91, 91, -91, 91, -104, 104, -104, 104, -117, 117, -117, 117, 130, -130, 130, -130, 143, -143, 143, -143, 156, -156, 156, -156, 169, -169]
 301 lines have not changed
 non scroll: 1 packets: [(300, 11)]
scroll encoding took 7ms
non-scroll encoding took 1ms
```
* "-d draw" on the client (added in [r14055](../commit/aa9393aa023e2aa14d2a7a8795d06e46d690ba4a), much less detailed than "-d paint") will show all the draw packets received, which should include some "scroll" encoding when the server prints "encode_scrolling(..)". Note however that scrolling detection is very efficient at skipping unchanged screen updates (ie: "301 lines have not changed" in the log just above), so in many cases there will not be any scroll packets at all. "scroll" draw packets are usually accompanied by lossless packets containing the other screen updates (ie: the scrollbar unless you also use #1295), ie when scrolling with firefox, you will get a series of about 10 updates like this one for each mouse wheel movement as it does its smooth scrolling repaint in small increments:
```
process_draw:       3   arrays for window   2 using scroll encoding at         (0, 0, 737, 557) with options={'av-sync': True, 'flush': 3}
process_draw:    2037    bytes for window   2 using    png encoding at        (0, 412, 737, 11) with options={'av-sync': True, 'flush': 2}
process_draw:    1510    bytes for window   2 using    png encoding at        (0, 495, 737, 11) with options={'av-sync': True, 'flush': 1}
process_draw:    1341    bytes for window   2 using    png encoding at         (0, 551, 737, 6) with options={'av-sync': True}
```
The scroll encoding contains the motion data, the png packets contain the area scrolled into view (the last packet) and the areas that have changed (ie: the top and bottom of the scrollbar which has also moved)

One last word: although the scroll paint code will use the stippled brown as "paint box colour", you can barely see it because we usually also paint the lossless screen updates together (flush>0) and those will include the same edges.. [r14058](../commit/51f5e7804a180f414aa5fa8fe44bdba2b69968a1) makes it 2 pixels thicker.

**Comment by totaam** [comment]
* the memleak is fixed in [r14107](../commit/953dc755e711ad564aca3ccce59ebe851deb45d2)
* moving quickly ("extreme corner case") is meant to use h264 since that's the best encoding there is for fast screen refreshes with no obvious scrolling detection possible.

**Comment by totaam** [comment]
Thanks!

Closing, we'll revisit this as part of #1341.

**Comment by totaam** [comment]
Re-opened because [r14382](../commit/47a3da5209ff894be94144b858240c9881698816) + [r14383](../commit/b743abf40be925c054a8a46ae0191aa7dc8a6523) make some changes to this core part of the code.

The big problem is that we have to spot when "video" (or scrolling in this case) stops and use a different encoding. Doing this detection is expensive to do, and hard to get right too.
The auto-refresh should fix the lossy screen update, does it not?

**Comment by totaam** [comment]
#1352: it seems that there are some painting issues.

**Comment by totaam** [comment]
Confirmed and fairly easy to reproduce.
[r14465](../commit/6abfae00757cfb30b95aedc43b0e08a9c6535b4f) improves the code and adds `XPRA_OPENGL_SAVE_BUFFERS` to the opengl client, making it possible to save the state of the client-side FBO, frame by frame.

Other TODO:
* `scroll data size mismatch: 876x1027 vs 876x1031` - we could trim or pad the checksum array

**Comment by totaam** [comment]
More changes in [r14466](../commit/32b1343d90b1ab21c7fd058ffcc46f5fee5e20ff), see commit message.
Using the newly introduced `XPRA_SAVE_VIDEO_FRAMES=jpeg` server side switch and the existing `XPRA_OPENGL_SAVE_BUFFERS=jpeg` client side switch.

How to use this cool new feature (should go on the wiki somewhere) for debugging any video picture issues:
* enable both env vars
* enable "compress" debug logging (and more? "scroll", whatever)
* wait for problem to occur
* inspect the recent sequence of images until we see the exact point where things don't look right
* grab the log sample around that point in time (the image filenames use the same format as the log timestamp)

----

Back to the bug: [r14467](../commit/7bc6d11dc08f68199e1c5b2f18b249486c89c4b6) fixes that, I can no longer reproduce it.
Note however that we just repaint the area, it may still appear out of order (looks like the picture is flickering a bit when it does), but the next screen refresh fixes it.

I'm still seeing the occasional lower than desired visual quality when scrolling very very fast. It should go back up more quickly than it does after you stop scrolling - but this is not really scrolling related and belongs in another ticket (related to #1257 and #1135).

It would be nice if we could also update the image checksum when we process the next auto-refresh (scrolling is currently considered lossy), but this would be quite hard to implement.

**Comment by totaam** [comment]
Let's finally close this.

**Comment by totaam** [comment]
Follow up in #1426 and #1320

**Comment by totaam** [comment]
See also #2248: more aggressive use of "scroll" encoding.

**Comment by totaam** [comment]
A bug has been fixed 3 years later! #2757
